File: Java/src/Linked List Cycle II.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 * int val;
 * ListNode next;
 * ListNode(int x) {
 * val = x;
 * next = null;
 * }
 * }
 */
public class Solution
{
  public ListNode detectCycle(ListNode head)
  {
    if (head == null || head.next == null)
      return null;
    ListNode fast = head;
    ListNode slow = head;

    while (fast != null)
    {
      slow = slow.next;
      fast = fast.next;
      if (fast == null)
        return null;
      fast = fast.next;
      if (fast == slow)
      {
        slow = head;
        while (slow != fast)
        {
          fast = fast.next;
          slow = slow.next;
        }
        return fast;
      }
    }
    return null;
  }
}

File: Python/Linked List Cycle II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    # @param head, a ListNode
    # @return a list node
    def detectCycle(self, head):
        if head == None: return None
        hare, turtle= head, head
        while hare != None:
            turtle = turtle.next
            hare = hare.next
            if hare == None: return None
            hare = hare.next
            if hare == turtle:
                turtle = head
                while turtle != hare:
                    hare = hare.next
                    turtle = turtle.next
                return hare
        return None